src/op_analytics/cli/subcommands/pulls/defillama/protocols.py

```diff
@@ -0,0 +1,185 @@
+from dataclasses import dataclass
+import polars as pl
+from op_analytics.coreutils.bigquery.write import (
+    most_recent_dates,
+    upsert_unpartitioned_table,
+)
+from op_analytics.coreutils.logger import structlog
+from op_analytics.coreutils.request import get_data, new_session
+from op_analytics.coreutils.threads import run_concurrently
+from op_analytics.coreutils.time import dt_fromepoch
+
+log = structlog.get_logger()
+
+PROTOCOLS_ENDPOINT = "https://api.llama.fi/protocols"
+PROTOCOL_DETAILS_ENDPOINT = "https://api.llama.fi/protocol/{slug}"
+
+BQ_DATASET = "uploads_api"
+PROTOCOL_METADATA_TABLE = "defillama_protocols_metadata"
+PROTOCOL_TVL_DATA_TABLE = "defillama_protocols_tvl"
+PROTOCOL_TOKEN_TVL_DATA_TABLE = "defillama_protocols_token_tvl"
+
+TVL_TABLE_LAST_N_DAYS = 7
+
+
+@dataclass
+class DefillamaProtocols:
+    """Metadata and data for all protocols."""
+
+    metadata_df: pl.DataFrame
+    app_tvl_df: pl.DataFrame
+    app_token_tvl_df: pl.DataFrame
+
+
+def pull_protocol_tvl(pull_protocols: list[str] | None = None) -> DefillamaProtocols:
+    """
+    Pulls and processes protocol data from DeFiLlama.
+
+    Args:
+        pull_protocols: list of protocol slugs to process. Defaults to None (process all).
+    """
+    session = new_session()
+
+    # Fetch the list of protocols and their metadata
+    protocols = get_data(session, PROTOCOLS_ENDPOINT)
+    metadata_df = extract_protocol_metadata(protocols)
+
+    # Construct URLs for protocol-specific data
+    urls = construct_urls(metadata_df, pull_protocols, PROTOCOL_DETAILS_ENDPOINT)
+    # Fetch protocol details in parallel
+    protocol_data = run_concurrently(lambda url: get_data(session, url), urls, max_workers=4)
+
+    # Extract protocol data into a dataframe
+    app_tvl_df, app_token_tvl_df = extract_protocol_tvl_to_dataframes(protocol_data)
+
+    upsert_unpartitioned_table(
+        df=metadata_df,
+        dataset=BQ_DATASET,
+        table_name=PROTOCOL_METADATA_TABLE,
+        unique_keys=["protocol_slug"],
+        create_if_not_exists=False,  # Set to True on first run
+    )
+
+    upsert_unpartitioned_table(
+        df=most_recent_dates(app_tvl_df, n_dates=TVL_TABLE_LAST_N_DAYS, date_column="date"),
+        dataset=BQ_DATASET,
+        table_name=PROTOCOL_TVL_DATA_TABLE,
+        unique_keys=["protocol_slug", "chain", "date"],
+        create_if_not_exists=False,  # Set to True on first run
+    )
+
+    upsert_unpartitioned_table(
```

Comment:
I remember you mentioning that some of these tables would need to be partitioned for the protocol level data

Reply:
sounds good, let's go with unpartitioned then